### DISCUSSION TITLE
QCoDeS Instrument Serialization

### DISCUSSION THREAD

**Discussion by estoneman** [microsoft/Qcodes#5707] [Q&A]
Hello, I am currently working on a project where it is required that an instance of a QCoDeS instrument class be shared across Python runtimes in some way (currently configured as a client/server architecture). My first attempt was to use a library like pickle in order to serialize/deserialize the instrument instance; however, pickling instances are explicitly unsupported as stated in many locations in the documentation like [here](https://microsoft.github.io/Qcodes/api/instrument/index.html#qcodes.instrument.IPInstrument.\_\_getstate\_\_). Is there a different way to capture the state of a live instrument object that can then be transmitted to another host?

**Comment by paul-lehmann-unibas** [comment]
I don't know if this meets your requirements, but you could proxy your instrument using proxying libraries (pyro or rpyc).
This way you could have one python instance (the server) run the qcodes drivers, which will talk to the actual instruments. And have several clients access the same driver object on the server. Proxying like this also has the advantage, that you have only one object and you do not need to keep several objects in sync.
Potential libraries would be:
- pyro5
- rpyc
- qcodes intstruementserver from toolsforexperiments. I don't know how actively it is maintained though.

**Reply by astafan8** [reply]
> qcodes intstruementserver from toolsforexperiments. I don't know how actively it is maintained though.

I also don't know, but it is likely in use. Better to ask the authors.

> Potential libraries would be:

Would you or your peers be interested in prototyping a solution? And open a PR with it and we can continue the discussion?

**Reply by paul-lehmann-unibas** [reply]
Unfortunately, I don't have time to work on something that's not of immediate use to our experiments, sorry.
Also I think I am not following completely. I did not want to suggest to add anything to qcodes, but rather suggest a general way to access objects in other python processes.
What did you have in mind?
